### ISSUE TITLE
Unsupported OS error when using Puppet Bolt

### ISSUE THREAD

**Issue by danifr** [OSC/puppet-module-openondemand#26]
I am trying to install OoD on a RHEL7 box using puppet bolt, however I am getting the following error:

 `Apply failed to compile for localhost: Unsupported OS: RedHat, module openondemand only supports RedHat 7 and 8 (file: /root/.puppetlabs/bolt/modules/openondemand/manifests/init.pp, line: 276, column: 5)`

To be honest, I am not sure if there is a bug in how bolt gets the facts or not.
If I list the OS facts from bolt I get:

```{family => RedHat, hardware => x86_64, selinux => {current_mode => enforcing, enabled => true, enforced => true, policy_version => 31, config_mode => enforcing, config_policy => targeted}, name => RedHat, architecture => x86_64, distro => {id => RedHatEnterpriseServer, codename => Maipo, description => Red Hat Enterprise Linux Server release 7.8 (Maipo), release => {full => 7.8, major => 7, minor => 8}, specification => :core-4.1-amd64:core-4.1-noarch}, release => {full => 7.8, major => 7.8}}```

As you can see (last line) `[OS][RELEASE][MAJOR]` is "7.8" instead of what it should be: "7".

As a workaround, do you see an issue if the module uses `[OS][DISTRO][RELEASE][MAJOR]` instead of `[OS][RELEASE][MAJOR]` ? They are both core facts: https://puppet.com/docs/puppet/latest/core_facts.html#os

I can prepare a PR for this.

**Comment by treydock** [comment]
I think making it look at either `os.release.major` or `os.distro.release.major` would work, so using  some kind of logic to pick one and if that's not present the other:

```puppet
$osmajor = $facts.dig('os', 'release', 'major')
if not $osmajor {
  $_osmajor = $facts.dig('os', 'distro', 'release', 'major')
} else {
  $_osmajor = $osmajor
}
...
$os = "${osfamily}-${_osmajor}"
```

I'm not sure that's a bolt issue, could be something unique to your system where facter uses different major OS version.

**Comment by danifr** [comment]
Thanks for your answer. Your patch wouldn't work in my case.

`$facts.dig('os', 'release', 'major')` exists and resolves to "7.8".

This is why I was suggesting simply assign `$osmajor` to `$facts.dig('os', 'distro', 'release', 'major')` which at least for me seems to have the right value.

**Comment by treydock** [comment]
Ah you're right. I think maybe the right solution is this then:

```puppet
  $osfamily = $facts.dig('os', 'family')
  $osmajor = $facts.dig('os', 'release', 'major')
  if $osfamily != 'RedHat' and $osmajor and versioncmp($osmajor, '7') < 0 {
...
```

The `versioncmp` should evaluate to `1` in your case. Might be better to change the `osmajor` to use `os.release.full` and make comparison to `7.0`.  I am hesitant to rely on a different fact as I'm not sure under what circumstance the `os.distro` fact is populated.  On my RHEL systems it's populated but not sure that's always the case. I know some facts only get populated with lsb packages installed and want to avoid those facts.

**Comment by danifr** [comment]
I understand your concern. 

Unfortunately, in my case I still need to get the major version number from somewhere. The module relies on the version number for configuring other things such as the repo url:

https://github.com/OSC/puppet-module-openondemand/blob/master/manifests/init.pp#L291

What about adding some code that checks:
```puppet
  $osmajor = $facts.dig('os', 'release', 'major')
  if $osmajor looks like X.Y
     $osmajor = X
  ...
  # and we keep the original check in place
  $supported = ['RedHat-7','RedHat-8']
  $os = "${osfamily}-${osmajor}"
  if ! ($os in $supported) {
    fail("Unsupported OS: ${osfamily}, module ${module_name} only supports RedHat 7 and 8. ($osmajor)")
  }
...

**Comment by treydock** [comment]
Ya that would work. I think you could use split regardless of what $osmajor looks like, using `split('7')[0]` will still produce 7, I think.

```
$osmajor = $facts.dig('os', 'release', 'major')
if $osmajor {
  $_osmajor = split($osmajor, '[.]')[0]
} else {
  $_osmajor = 'Unknown'
}
```

**Comment by treydock** [comment]
The fix for this is in v0.9.2 via #27
